perf: optimizing theme config loading

h1k3r opened the pull request:
### 1. Why is this change necessary?

In different places of the system `StorefrontPluginRegistry::getConfigurations` or `StorefrontPluginRegistry::getByTechnicalName` are used to load theme config. That leads to all theme.json files and all other storefront configs being parsed & generated on the fly just to access some information information from the config for the current theme.

### 2. What does this change do, exactly?

This change introduces a new database table `theme_runtime_config` and relevant object `ThemeRuntimeConfig` to store precalculated configs that should be generated during `refreshTheme` call. It should be used in all runtime calls instead of `StorefrontPluginRegistry`

### 3. Describe each step to reproduce the issue or behavior.


### 4. Please link to the relevant issues (if any).
- Closes https://github.com/shopware/shopware/issues/7766
- Partial fix was introduced in https://github.com/shopware/shopware/commit/846c6aeaad0c2914a44b05ba1a65a0a18dc00b19 

### 5. Checklist

- [ ] I have written tests and verified that they fail without my change
- [ ] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [ ] I have read the contribution requirements and fulfill them.

### 6. What to test

- [x] Plugin theme
- [x] App theme
- [x] Theme configuration field that is also used as css variable.
- [x] Theme configuration field with the same name (override).
- [x] Image theme configuration field
- [x] Changing the theme configuration in the admin.
- [x] Theme inheritance (both parent and child add theme config fields)
- [ ] Theme config inheritance https://developer.shopware.com/docs/guides/plugins/themes/theme-inheritance-configuration.html
- [x] Copied themes (copy functionality in administration)
- [ ] Test theme compilation without DB: https://developer.shopware.com/docs/guides/hosting/installation-updates/deployments/build-w-o-db.html
- [x] Test "@StorefrontBootstrap" placeholder for CSS.
- [ ] Test if adding/changing the asset path is possible and assets can be loaded as documented.
- [ ] Test adding SCSS variables via plugin config: https://developer.shopware.com/docs/guides/plugins/plugins/storefront/add-scss-variables.html 
- [ ] Shop update to the version with a change (non-default theme should be configured)
- [ ] Snippets rendering (no inactive themes snippets, snippets rendering in rufus)

h1k3r committed:
Added ThemeRuntimeConfig, switched to it in icons sets preparation in TemplateDataSubscriber

h1k3r committed:
Changed ThemeInheritanceBuilder to use ThemeRuntimeConfig (required database)

h1k3r committed:
Changed ResolvedConfigLoader to use ThemeRuntimeConfig

h1k3r committed:
Changed SnippetService to use ThemeRuntimeConfig

h1k3r committed:
Changed ThemeScripts to use ThemeRuntimeConfig

aragon999 commented:
I like the idea in general, but I am wondering why not putting it into a "normal" cache.

This would have the advantage that this information could also be stored in a redis db or so?

h1k3r committed:
Migration test

h1k3r committed:
Added ThemeRuntimeConfigTest

h1k3r committed:
Added ThemeScriptsTest, improved ThemeScripts readability

h1k3r commented on src/Storefront/Theme/ThemeScripts.php:
```diff
@@ -41,34 +38,31 @@ public function getThemeScripts(): array
             return [];
         }
 
-        $themeName = $request->attributes->get(SalesChannelRequest::ATTRIBUTE_THEME_NAME, SalesChannelRequest::ATTRIBUTE_THEME_BASE_NAME)
```
`SalesChannelRequest::ATTRIBUTE_THEME_BASE_NAME` was providing `theme-name` as default value for theme name here. It looked like mistake.

Also, since https://github.com/shopware/shopware/commit/12dc4a242a00b59bdf4fe668d0022e6f3f7393db we do not load config by name, but rather by id. That means loading parent theme configuration does not work.

With changes in this pr we do not do any calculation in realtime, so probably all base theme logic can be removed 🤔

mitelg replied:
wasn't there an issue with theme config inheritance :thinking:

shyim replied:
yes you needed this base name stuff, when you made a child theme in the admin UI. attribute theme name is then not filled

h1k3r replied:
But now we use only theme config to load files. So I suppose the change is ok.

keulinho replied:
> Copied themes (copy functionality in administration)


That is in the test manual, so when it still works as expected we are good right?

h1k3r replied:
Removed the check completely.

h1k3r commented on src/Storefront/Theme/ThemeLifecycleService.php:
```diff
@@ -115,6 +116,9 @@ public function refreshTheme(StorefrontPluginConfiguration $configuration, Conte
         $this->themeChildRepository->delete($toDeleteIds, $context);
         $this->themeChildRepository->upsert($parentThemes, $context);
+
+        // we don't resolve files as theme can be refreshed before it's built
```
Currently runtime configuration is built during refresh stage of theme lifecycle. It can happen that theme is not built at this stage, what makes it impossible to resolve theme files (at least without bigger changes to underlying processes). So currently files are resolved on the first call to the `ThemeRuntimeConfigService::getResolvedRuntimeConfig`, and whole config is recalculated.

This is a temporary solution that must be changed!

h1k3r commented on src/Core/System/Snippet/SnippetService.php:
```diff
@@ -50,6 +50,7 @@ public function __construct(
          */
         private readonly ContainerInterface $container,
         private readonly ExtensionDispatcher $extensionDispatcher,
+        private readonly ThemeRuntimeConfigService $themeRuntimeConfigService,
```
Should this be an optional dependency, like $salesChannelThemeLoader?

mitelg replied:
yes, we cannot have classes from Storefront as dependencies in the core

h1k3r commented:
Hi @aragon999 ,
thank you for your question.

> I like the idea in general, but I am wondering why not putting it into a "normal" cache.
> This would have the advantage that this information could also be stored in a redis db or so?

As data in config is not dynamic and can be computed only when theme is updated/configurations saved into db, mysql as persistent storage looked like a good option. Also not every Shopware installation has Redis available, but we would like to have this optimization working for everyone.

Sure caching into Redis can be added later for a better performance later.

h1k3r committed:
fix for snippet service tests, made dependency on storefront components optional

h1k3r committed:
allowing theme_runtime_config to have no definition

shyim commented on src/Core/Migration/V6_7/Migration1742568836CreateThemeRuntimeConfigTable.php:
```diff
@@ -0,0 +1,35 @@
+    public function update(Connection $connection): void
+    {
+        $connection->executeStatement(<<<'SQL'
+            CREATE TABLE IF NOT EXISTS `theme_runtime_config` (
```
I think `technical_name` requires an index, you filter by it

h1k3r replied:
Done 👍

shyim commented on src/Storefront/Theme/ThemeRuntimeConfigService.php:
```diff
@@ -0,0 +1,191 @@
+ * @internal
+ */
+#[Package('framework')]
+class ThemeRuntimeConfigService
```
we do just one query per request right?

shyim replied:
because if we need here some caching,

h1k3r replied:
Added caching into the local variable

shyim commented on src/Storefront/Theme/ThemeScripts.php:
```diff
@@ -41,34 +38,31 @@ public function getThemeScripts(): array
-            $themeConfig = $this->configLoader->load($themeId, $salesChannelContext->getContext());
+        $path = $this->themePathBuilder->assemblePath($salesChannelId, $themeId);
+
+        return $this->cache->get('theme_scripts_' . $path, function (ItemInterface $item) use ($themeId) {
```
do we really need this cache here anymore 🤔, when the query is simple maybe we can skip this

h1k3r replied:
removed

h1k3r committed:
fixed type hinting for ThemeRuntimeConfig

h1k3r committed:
fixed ThemeInheritanceBuilder tests

h1k3r committed:
fixed theme refresh during installation/in some tests (configs collection should be passed from ThemeLifecycleHandler)

h1k3r committed:
fixed minor bugs, phpstan errors, use domain exceptions

h1k3r committed:
removed cache invalidation, as caching using theme_scripts_ key was removed previously

h1k3r committed:
deprecated CachedResolvedConfigLoader

h1k3r commented on src/Storefront/Theme/CachedResolvedConfigLoaderInvalidator.php:
```diff
@@ -49,14 +49,19 @@ public function assigned(ThemeAssignedEvent $event): void
+        $this->cacheInvalidator->invalidate([self::buildCacheTag($event->getThemeId())]);
+    }
+
+    public static function buildCacheTag(string $themeId): string
```
Should we rename whole class to `ConfigCacheInvalidator`? We invalidate by generated tag here, in `ThemeCompiler`, and add tags in `ThemeConfigValueAccessor`

mitelg replied:
yes :+1:

keulinho replied:
Maybe `ThemeConfigCacheInvalidator` -> to not confuse this with invalidating system configs or any other config?

I know `Theme` is already in the namespace, however you see that only in the import statement, not in the directly usages of that class

h1k3r replied:
Done

h1k3r committed:
refactored ThemeRuntimeConfigService

-using proper exception
-caching retrieved data in local variables
-extracted working with db into a separate class

h1k3r committed:
Added few tests

h1k3r committed:
Added missing index

h1k3r committed:
Renamed runtime config refresh method, added method to clean caches, fixed MigrationTest to avoid table deletion

h1k3r committed:
fixing rufus test error related to caching, rollback test changes

h1k3r committed:
added changelog

h1k3r committed:
fixed pipelines, added additional tests

h1k3r committed:
Merge branch 'trunk' into optimizing-theme-config-loading

h1k3r committed:
added runtime config updated to the theme compilation and configuration updates

h1k3r committed:
added runtime config updated to the theme compilation and configuration updates

h1k3r committed:
catching more exceptions types in files resolution

h1k3r committed:
-made technicalName nullable, added support for theme copies

h1k3r committed:
-added runtime config generation for the shopware upgrade case

h1k3r committed:
Small optimization, additional tests

h1k3r committed:
Merge branch 'trunk' into optimizing-theme-config-loading

mitelg commented on changelog/_unreleased/2025-04-09-optimizing-theme-config-loading.md:
```diff
@@ -0,0 +1,25 @@
+___
+# Next Major Version Changes
+## Removal of CachedResolvedConfigLoader
+* The `\Shopware\Storefront\Theme\CachedResolvedConfigLoader` will be removed.
```
```suggestion
* The `\Shopware\Storefront\Theme\CachedResolvedConfigLoader` was removed.
```

this changelog will be part of 6.8, were this removal already happend

h1k3r replied:
Done

mitelg commented on src/Core/System/Snippet/SnippetService.php:
```diff
@@ -259,9 +260,12 @@ protected function getUnusedThemes(array $usingThemes = []): array
             return [];
```
this condition can be removed.

can the method `getUsedThemes` also be refactored? if so the injection of the DI container could probably be removed, too.

shyim replied:
I think it would be better when we would fire an event and listen to that event in Storefront Bundle. It's messy to use Storefront classes in Core namespace

BrocksiNet replied:
I agree ⬆️, but why should we do this now in this PR?

Let's move this to a new issue. 🌞

mitelg replied:
because @h1k3r is introducing another storefront dependency. why should we not do it correctly right now?

h1k3r replied:
Done, please have another look

h1k3r committed:
Fixed test coverage attribute, removed unneeded attribute reading in ThemeScripts

mitelg commented on src/Storefront/Theme/Exception/ThemeException.php:
```diff
@@ -73,4 +75,45 @@ public static function themeCompileException(string $themeName, string $message
+        array $childThemeSalesChannel,
+        array $assignedSalesChannels,
+        ?\Throwable $e = null,
+    ): ThemeAssignmentException {
```
~ThemeAssignmentException should then extend from ThemeException~

we do not catch this exception ourselves, so it should be deprecated

h1k3r replied:
done

mitelg commented on src/Storefront/Theme/ThemeLifecycleService.php:
```diff
@@ -64,11 +65,14 @@ public function refreshThemes(
+    /**
+     * @deprecated tag:v6.8.0 parameter $configurationCollection will be added - reason:new-optional-parameter
+     */
+    public function refreshTheme(StorefrontPluginConfiguration $configuration, Context $context/* , ?StorefrontPluginConfigurationCollection $configurationCollection = null */): void
```
do we need to comment the parameter out at all? :thinking: as the service has no interface, it shouldn't be breaking due to the default value

h1k3r replied:
bc checker complained 🙈

shyim replied:
can we mark the class then as final for next major?

h1k3r replied:
Done

mitelg commented on src/Storefront/Theme/ThemeLifecycleService.php:
```diff
@@ -115,6 +119,17 @@ public function refreshTheme(StorefrontPluginConfiguration $configuration, Conte
+        $configurationCollection = \func_num_args() === 3 ? \func_get_arg(2) : null;
+
+        // we don't resolve files as theme can be refreshed before it's built
+        $filesRequired = false;
```
this variable is never changed, so it can directly set on the method call

h1k3r replied:
I added variable to have clear what comment describes. It's a tricky part, wanted to have intention described. Do you still think it makes sense to remove it?

mitelg replied:
no, should be fine

mitelg commented on src/Storefront/Theme/ThemeMergedConfigBuilder.php:
```diff
@@ -0,0 +1,404 @@
+     * @throws ThemeException
+     * @throws InconsistentCriteriaIdsException
+     *
+     * @return array<string, mixed>
```
can be introduce a DTO instead?

h1k3r replied:
I do not want to include this into the scope of the already long-running task. I just extracted a part of existing functionality from ThemeService into this class (without any changes) to resolve circular dependency issue. Anyway can try to do it if you insist.

mitelg replied:
if this is too much effort and just old code moved around, we can leave it for now. 

but for new code we should use DTOs in such cases instead

mitelg commented on src/Storefront/Theme/ThemeRuntimeConfigService.php:
```diff
@@ -0,0 +1,260 @@
+    /**
+     * Refreshes the whole ThemeRuntimeConfig object.
+     */
+    public function refreshRuntimeConfig(string $themeId, StorefrontPluginConfiguration $themeConfig, Context $context, bool $failOnFileResolveError = false, ?StorefrontPluginConfigurationCollection $configCollection = null): ThemeRuntimeConfig
```
a line wrap would be nice here :grin:

mitelg commented on src/Storefront/Theme/ThemeRuntimeConfigService.php:
```diff
@@ -0,0 +1,260 @@
+        }
+    }
+
+    public function resetCaches(): void
```
this seems to be unused :thinking: did you forget to implement the ResetInterface?

h1k3r replied:
It's used in the `\Shopware\Storefront\Theme\ThemeLifecycleService::refreshTheme`

mitelg commented on src/Storefront/Theme/ThemeScripts.php:
```diff
@@ -41,7 +34,7 @@ public function getThemeScripts(): array
         }
 
-        $themeName = $request->attributes->get(SalesChannelRequest::ATTRIBUTE_THEME_NAME, SalesChannelRequest::ATTRIBUTE_THEME_BASE_NAME)
+        $themeName = $request->attributes->get(SalesChannelRequest::ATTRIBUTE_THEME_NAME)
```
is the name needed at all now?

h1k3r replied:
That's a good question. Formally if we remove it - we change the existing behavior (before - no scripts if attribute is missing, after - render scripts), but I am not sure if there are reasonable scenarios when this can happen.

mitelg commented on src/Storefront/Theme/ThemeService.php:
```diff
@@ -197,130 +218,15 @@ public function resetTheme(string $themeId, Context $context): void
      */
     public function getThemeConfiguration(string $themeId, bool $translate, Context $context): array
```
should we deprecate those methods? as they now only call another service?

h1k3r replied:
The idea was to keep `ThemeService` as public interface, `MergedConfigBuilder` as internal implementation, that should not be exposed (and may have other methods added).
Or can we totally internalize these methods? Wdyt?

mitelg replied:
okay, if MergedConfigBuilder is internal, we should keep those public methods then :+1:

mitelg commented on tests/unit/Storefront/Theme/ThemeMergedConfigBuilderTest.php:
```diff
@@ -0,0 +1,1016 @@
+        $parentThemeId = Uuid::randomHex();
+        $baseThemeId = Uuid::randomHex();
+
+        return [
```
such big arrays will slow down PHPStan. could you extract that into a separate file?

h1k3r replied:
I am not sure it makes sense to move this array to some place not covered by phpstan. It is nice to have checks for the structure, especially taking into account that parts of it are generated dynamically. I've replaced one array with yields for every test case (also this is an old code moved from another place during small refactoring, so no additional slowdown introduced).

mitelg replied:
you do not exclude something from PHPStan. just moving it to another file. you can see examples here https://github.com/shopware/shopware/pull/6528/files

h1k3r replied:
Done (but not sure why it should make phpstan faster, if it still parses the array)

mitelg replied:
don't ask me about the internal of PHPStan, but it has quite an impact, if such arrays are separated from already big files :man_shrugging:

mitelg commented on tests/unit/Storefront/Theme/ThemeRuntimeConfigServiceTest.php:
```diff
@@ -0,0 +1,399 @@
+
+        $result = $this->service->refreshRuntimeConfig($themeId, $themeConfig, $context, $filesRequired, $configCollection);
+
+        static::assertEquals($themeId, $result->themeId);
```
please use `assertSame` where possible instead, as it is type safe

h1k3r replied:
Updated for most of cases (had to keep in some old tests)

keulinho commented on src/Storefront/Theme/ThemeRuntimeConfig.php:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    /**
+     * @param array{
```
does it make sense to add this as a local type alias? so we don't have to redefine the same type information all over the place? https://phpstan.org/writing-php-code/phpdoc-types#local-type-aliases

I expect this is used more often, as that is also the structure we get back from the DB right?

h1k3r replied:
Actually it's used only in this place. Method `with` has all fields optional. But I moved it into the local type alias.

keulinho commented on tests/integration/Storefront/Theme/ThemeInheritanceBuilderTest.php:
```diff
@@ -185,3 +180,44 @@ private function getMockedPlugin(string $pluginName, string $pluginClass): Bundl
+/**
+ * @internal
+ */
+class ThemeRuntimeConfigTestService extends ThemeRuntimeConfigService
```
Does it make sense to put this under src/Test/Stub namespace, similiar what we have in the core namespace for staticEntityRepo, etc? Because for me this looks like this stub will be handy in every unit test where you test a class that collaborates with the themeRuntimeService

h1k3r replied:
Done

keulinho commented on tests/integration/Storefront/Theme/ThemeInheritanceBuilderTest.php:
```diff
@@ -185,3 +180,44 @@ private function getMockedPlugin(string $pluginName, string $pluginClass): Bundl
+
+    public function initFromConfigCollection(StorefrontPluginConfigurationCollection $configurationCollection): void
+    {
+        $this->configs = [];
```
Mhm i don't like that this method carries state over between test cases 🤔 

In unit tests I try to create the classes from scratch in every test, creating an instance once in the setup is nice for integration tests 

Any specific reason for this choice? Because I see the risk that suddenly tests become independent of execution order, which could be unexpected and hard to debug when you don't call initFromConfiguration on every test case 🤔 

so what I had in mind would be to do the initialising in the constructor and then always create a new instance in the test methods

What do you think?

h1k3r replied:
Good point!
Current code, I think, is a result of "anchoring" to the old structure. Previously we had a mock, which we configured in every test, I replaced mock with the service.
It's easy to introduce the factory method that will create builder with configs provided.

h1k3r replied:
Replaced with initialization in constructor.

keulinho commented on tests/unit/Core/System/Snippet/SnippetServiceTest.php:
```diff
@@ -74,19 +73,8 @@ public function testGetStorefrontSnippets(
-            $themeRegistry->expects($this->once())->method('getConfigurations')->willReturn($plugins);
-            $container->expects($this->once())->method('get')->with(StorefrontPluginRegistry::class)->willReturn($themeRegistry);
-        }
+        $runtimeConfigService = $this->createMock(ThemeRuntimeConfigService::class);
```
Here (and in the next test as well, not sure if in more as well) you could re-use the stub implementation from the previous test case, in some cases this might be more code to set it up (but we might create easy factory methods where you need to pass the names and the stub will create the correct config objects)

This has the benefit that the test is more stable when we refactor the implementation and call different methods etc (for context https://dev.to/shopware/mocks-and-their-influence-on-software-design-3o49)

h1k3r replied:
Added factory methods, also in some other tests.

keulinho commented on tests/unit/Storefront/Theme/ThemeRuntimeConfigStorageTest.php:
```diff
@@ -0,0 +1,305 @@
+ */
+#[Package('framework')]
+#[CoversClass(ThemeRuntimeConfigStorage::class)]
+class ThemeRuntimeConfigStorageTest extends TestCase
```
should we rather move this to integration test? 

IMHO mocking SQL calls in tests is quite useless, I'll rather test the SQL for real and then put a coverage ignore on the class

you already encapsulated the storage to it's own class, so the rest of the business logic can be tested in isolation in pure unit way -> that is really cool

Speaking from experience of doing a DBAL major update (I guess you experienced something similar ;) ) those tests are really hard to update when DBAL changes something

nfortier-shopware replied:
I agree with this approach, mocking SQL comes handy only when you need to skip db connection and continue to test further logic within a class. When the class itself is pure isolated SQL involvement then integration is the way

h1k3r replied:
Done

nfortier-shopware commented on tests/unit/Storefront/Theme/ThemeRuntimeConfigServiceTest.php:
```diff
@@ -0,0 +1,399 @@
+        yield 'config found' => [
+            '1234567890abcdef1234567890abcdef',
+            'test-theme',
+            self::createThemeRuntimeConfig(),
```
let's be more explicit with:
```
yield 'no record found' => [
            'themeId' => '1234567890abcdef1234567890abcde1',
            'technicalName' => 'nonexistent-theme-name',
            'expectedConfig'  => null,
        ];
...
```

this will prevent to have mixtures in providers of named/unnamed parameters and it's more readable

h1k3r replied:
Done

nfortier-shopware commented on tests/unit/Storefront/Theme/ThemeMergedConfigBuilderTest.php:
```diff
@@ -0,0 +1,1016 @@
+        );
+
+        $this->expectException(ThemeException::class);
+        $this->expectExceptionMessage(\sprintf('Could not find theme with id "%s"', $themeId));
```
`expectExceptionObject` would be better

h1k3r replied:
Done

nfortier-shopware commented on tests/unit/Storefront/Theme/ThemeMergedConfigBuilderTest.php:
```diff
@@ -0,0 +1,1016 @@
+        static::assertArrayHasKey('fields', $config);
+        static::assertArrayHasKey('currentFields', $config);
+        static::assertArrayHasKey('baseThemeFields', $config);
+        static::assertEquals($expected, $config);
```
static::assertSame() would be better

h1k3r replied:
This test was moved from other file. Changing to assertSame would require to change testing data, which I would not do in the scope of such big change (it may be hard to understand if the data was changed correctly or new bug was introduced).

nfortier-shopware commented on tests/unit/Storefront/Theme/ThemeRuntimeConfigServiceTest.php:
```diff
@@ -0,0 +1,399 @@
+        $serviceMock = $this->createPartialMock(
+            ThemeRuntimeConfigService::class,
+            ['refreshRuntimeConfig', 'getRuntimeConfig'] // Only mock this method
+        );
```
createPartialMock should be avoided as much as possible (there is only one place in platform we use this and about to be removed,) especially when getResolvedRuntimeConfig method is public. Is there something blocking us to test it without?

h1k3r replied:
Done. You may want to have another look.

h1k3r committed:
PR comments fixes

h1k3r committed:
PR comments fixes for tests, missing changelog records

h1k3r committed:
Replaced unit test for ThemeRuntimeConfigStorage with integration

h1k3r committed:
Renamed CachedResolvedConfigLoaderInvalidator to ThemeConfigCacheInvalidator

h1k3r committed:
Merge branch 'trunk' into optimizing-theme-config-loading

h1k3r committed:
Deprecated ThemeAssignmentException

h1k3r committed:
stylefixes after merge

h1k3r committed:
fixed ThemeRuntimeConfigStorageTest dependency on other tests

h1k3r committed:
moved big array from ThemeMergedConfigBuilderTest into fixture class

h1k3r committed:
removed calculation of unused theme name parameter from ThemeScripts

h1k3r committed:
decoupled SnippetService from dependencies from the Storefront namespace

h1k3r committed:
Merge branch 'trunk' into optimizing-theme-config-loading

mitelg commented on changelog/_unreleased/2025-04-09-optimizing-theme-config-loading.md:
```diff
@@ -0,0 +1,37 @@
+issue: https://github.com/shopware/shopware/issues/7766
+---
+# Core
+* Changed `\Shopware\Core\System\Snippet\SnippetService` to use `\Shopware\Storefront\Theme\ThemeRuntimeConfigService` to get information about available themes
```
is this still correct, as you have implemented an event?

h1k3r replied:
fixed

mitelg commented on changelog/_unreleased/2025-04-09-optimizing-theme-config-loading.md:
```diff
@@ -0,0 +1,37 @@
+# Storefront
+* Added new `\Shopware\Storefront\Theme\ThemeRuntimeConfig` entity and `theme_runtime_config` table to store theme runtime configuration
+* Added `\Shopware\Storefront\Theme\ThemeRuntimeConfigService` to handle theme runtime configurations
+* Changed `\Shopware\Storefront\Theme\ThemeLifecycleService`, adding optional `$configurationCollection` parameter to the `refreshTheme` method, and deprecating class to be marked as final in the next major version.
```
```suggestion
* Changed `\Shopware\Storefront\Theme\ThemeLifecycleService`, adding optional `$configurationCollection` parameter to the `refreshTheme` method
* Deprecated `\Shopware\Storefront\Theme\ThemeLifecycleService` to be marked as final in the next major version.
```

mitelg commented on changelog/_unreleased/2025-04-09-optimizing-theme-config-loading.md:
```diff
@@ -0,0 +1,37 @@
+___
+# Upgrade Information
+## Theme configuration changes
+* Theme configuration used during storefront rendering are now stored in a `theme_runtime_config` table and regenerated on the refresh stage of theme lifecycle.
```
```suggestion
* Theme configuration used during storefront rendering is now stored in a `theme_runtime_config` table and regenerated on the refresh stage of theme lifecycle.
```

mitelg commented on changelog/_unreleased/2025-04-09-optimizing-theme-config-loading.md:
```diff
@@ -0,0 +1,37 @@
+* Changed `\Shopware\Storefront\Theme\CachedResolvedConfigLoaderInvalidator` name to `\Shopware\Storefront\Theme\ThemeConfigCacheInvalidator`
+* Deprecated `\Shopware\Storefront\Theme\CachedResolvedConfigLoader`, as it is no longer used in the storefront
+* Deprecated `\Shopware\Storefront\Theme\Exception\ThemeAssignmentException`
+* Deprecated `\Shopware\Core\System\Snippet\SnippetService::getUnusedThemes` method, replacement will not be provided.
```
maybe a small hint, how to get those data now? :thinking:

h1k3r replied:
Not sure that makes sense. It looks like this method was initially introduced for internal usage, new event was explicitly marked internal. But if someone needs that - they will be able to figure that out from changelog Core section.

mitelg replied:
ack :heavy_check_mark:

mitelg commented on changelog/_unreleased/2025-04-09-optimizing-theme-config-loading.md:
```diff
@@ -0,0 +1,37 @@
+* The `\Shopware\Storefront\Theme\Exception\ThemeAssignmentException` was removed.
+
+## Changes to ThemeLifecycleService
+* The `\Shopware\Storefront\Theme\ThemeLifecycleService` became final.
```
```suggestion
* The `\Shopware\Storefront\Theme\ThemeLifecycleService` is now final.
```

mitelg commented on src/Core/System/Snippet/Event/SnippetsThemeResolveEvent.php:
```diff
@@ -0,0 +1,67 @@
+    /**
+     * @var list<string>
+     */
+    private array $usedThemes = [];
```
why not public properties without getter/setter?

h1k3r replied:
The event is mutable. I like having setter to have more explicit/visible write operations. But I can change to public methods if you prefer.

mitelg replied:
I have no strong opinion on that. but we should really commit on something, as I saw both ways :see_no_evil:

keulinho replied:
i also don't mind either, not sure if it is worth to have common pattern for that, only thing for array we could add validation to the setter or offer `add` and `remove` functionality instead of overriding the whole array

mitelg replied:
recently added events have public properties, but from readability sight, it makes it more clearer to have a setter method, imo. 

maybe a distinction between readonly, just notifying events, and events, where you are able to mutate the properties?

h1k3r replied:
Ok, I keep it as it is for the moment. If we come up with a common standard - that can be easily changed (this event is marked internal).

mitelg commented on src/Storefront/Theme/Message/DeleteThemeFilesHandler.php:
```diff
@@ -40,8 +38,5 @@ public function __invoke(DeleteThemeFilesMessage $message): void
         }
 
         $this->filesystem->deleteDirectory('theme' . \DIRECTORY_SEPARATOR . $message->getThemePath());
-        $this->cacheInvalidator->invalidate([
```
the same needs to be done in `\Shopware\Storefront\Theme\ScheduledTask\DeleteThemeFilesTaskHandler`

h1k3r replied:
Done

h1k3r committed:
added ThemeSnippetsSubscriberTest, removed final from DatabaseSalesChannelThemeLoader to improve tests

mitelg commented on changelog/_unreleased/2025-04-09-optimizing-theme-config-loading.md:
```diff
@@ -0,0 +1,37 @@
+* The `\Shopware\Storefront\Theme\CachedResolvedConfigLoader` is now deprecated and will be removed in the next major version. Please update the code that directly uses it to use the `\Shopware\Storefront\Theme\ResolvedConfigLoader` instead.
+* The `\Shopware\Storefront\Theme\Exception\ThemeAssignmentException` is now deprecated and will be removed in the next major version. Please use `\Shopware\Storefront\Theme\Exception\ThemeException::themeAssignmentException`.
+___
+# Next Major Version Changes
```
we already did that on other PRs. would be nice, if you could put that in the 6.8 upgrade guide already

h1k3r replied:
@mitelg updated, thanks for the suggestion

mitelg commented on changelog/_unreleased/2025-04-09-optimizing-theme-config-loading.md:
```diff
@@ -0,0 +1,37 @@
+* Deprecated `\Shopware\Storefront\Theme\Exception\ThemeAssignmentException`
+* Deprecated `\Shopware\Core\System\Snippet\SnippetService::getUnusedThemes` method, replacement will not be provided.
+___
+# Upgrade Information
```
```suggestion
# Upgrade Information

```

this will prevent merge conflicts :grin:

h1k3r committed:
removed remaining theme scripts cache invalidation

keulinho commented on src/Core/System/Snippet/SnippetService.php:
```diff
@@ -249,21 +251,17 @@ public function findSnippetSetId(string $salesChannelId, string $languageId, str
      * @return list<string>
      */
-    protected function getUnusedThemes(array $usingThemes = []): array
+    protected function getUnusedThemes(array $usingThemes = []/* , array $unusedThemes */): array
```
that's a pretty creative way to handle the BC break, nice 👍

h1k3r committed:
-fixed possible bug with reading runtime config by name when theme is not compiled

h1k3r committed:
-removed usage of partial mock

h1k3r committed:
Apply suggestions from code review

Co-authored-by: Michael Telgmann <mitelg@users.noreply.github.com>

h1k3r committed:
-added description of new SnippetsThemeResolveEvent related changed to the changelog

h1k3r committed:
-moved record from changelog to upgrade guide

h1k3r committed:
Merge branch 'trunk' into optimizing-theme-config-loading

h1k3r committed:
Merge branch 'trunk' into optimizing-theme-config-loading

h1k3r committed:
Merge branch 'trunk' into optimizing-theme-config-loading

h1k3r committed:
Merge branch 'trunk' into optimizing-theme-config-loading

h1k3r committed:
Merge branch 'trunk' into optimizing-theme-config-loading

mitelg commented on src/Storefront/Resources/app/administration/package-lock.json:
as you didn't change the package.json, this should not be changed. could always cause side effects to update deps in JS stuff :grimacing:

h1k3r replied:
Thanks, accidentally committed during the merge

h1k3r committed:
Merge branch 'trunk' into optimizing-theme-config-loading

h1k3r committed:
remove accidentally added package-lock.json

h1k3r committed:
Merge branch 'trunk' into optimizing-theme-config-loading
